clippy_lints/src/unused_io_amount.rs

```diff
@@ -45,126 +46,241 @@ declare_clippy_lint! {
+                check_expr(cx, exp);
+            }
+        }
+
+        if let Some(exp) = block.expr {
+            match exp.kind {
+                hir::ExprKind::If(_, _, _) | hir::ExprKind::Match(_, _, _) => check_expr(cx, exp),
+                _ => {},
+            }
+        }
+    }
+}
+
+fn check_expr<'a>(cx: &LateContext<'a>, expr: &'a hir::Expr<'a>) {
+    match expr.kind {
+        hir::ExprKind::If(cond, _, _)
+            if let ExprKind::Let(hir::Let { pat, init, .. }) = cond.kind
+                && pattern_is_okay_underscore(cx, pat)
+                && let Some(op) = should_lint(cx, init) =>
+        {
+            emit_lint(cx, cond.span, op, &[pat.span]);
+        },
+        hir::ExprKind::Match(expr, arms, hir::MatchSource::Normal) if let Some(op) = should_lint(cx, expr) => {
+            let found_arms: Vec<_> = arms
+                .iter()
+                .filter_map(|arm| {
+                    if pattern_is_okay_underscore(cx, arm.pat) {
+                        Some(arm.span)
+                    } else {
+                        None
                     }
-                } else {
-                    check_map_error(cx, res, expr);
-                }
-            },
-            hir::ExprKind::MethodCall(path, arg_0, ..) => match path.ident.as_str() {
-                "expect" | "unwrap" | "unwrap_or" | "unwrap_or_else" | "is_ok" | "is_err" => {
-                    check_map_error(cx, arg_0, expr);
-                },
-                _ => (),
-            },
-            _ => (),
+                })
+                .collect();
+            if !found_arms.is_empty() {
+                emit_lint(cx, expr.span, op, found_arms.as_slice());
+            }
+        },
+        _ if let Some(op) = should_lint(cx, expr) => {
+            emit_lint(cx, expr.span, op, &[]);
+        },
+        _ => {},
+    };
+}
+
+fn should_lint<'a>(cx: &LateContext<'a>, mut inner: &'a hir::Expr<'a>) -> Option<IoOp> {
+    inner = unpack_match(inner);
+    inner = unpack_try(inner);
+    inner = unpack_call_chain(inner);
+    inner = unpack_await(inner);
+    // we type-check it to get whether it's a read/write or their vectorized forms
+    // and keep only the ones that are produce io amount
+    check_io_mode(cx, inner)
+}
+
+fn is_okay_pat(cx: &LateContext<'_>, pat: &hir::Pat<'_>) -> bool {
+    if let PatKind::TupleStruct(ref path, [inner_pat], ddp) = pat.kind
+        && ddp.as_opt_usize().is_none()
+        && is_res_lang_ctor(cx, cx.qpath_res(path, inner_pat.hir_id), hir::LangItem::ResultOk)
+    {
+        return true;
+    }
+    false
+}
+
+fn pattern_is_okay_underscore(cx: &LateContext<'_>, pat: &hir::Pat<'_>) -> bool {
+    is_okay_pat(cx, pat)
+        && matches!(
+            pat.kind,
+            PatKind::TupleStruct(
+                _,
+                [hir::Pat {
+                    kind: PatKind::Wild,
+                    ..
+                }],
+                _
+            )
+        )
+}
```

Comment:
Should these 2 move to `clippy_utils` ?

Reply:
As Alejandra pointed out on Zulip, there's also the case of `Ok(_@bind)`.

Reply:
I haven't yet fully looked through the code so may be missing context, but if the intention here is to check if the pattern ignores the bytes, then this shouldn't return true for `Ok(bind @ _)`. It's basically the same as `Ok(bind)`, which should be considered read. I think this is handled correctly right now though. Wouldn't hurt to have a test for that in any case

Reply:
It shouldn't because it is matching the wild card. If for some reason the user has a bind there, then either they get a lint about unused var - which is why I didn't add a check here for that case - or they handle it, so it should be cool. 

I will add the test case it tomorrow.

Reply:
> It shouldn't because it is matching the wild card

Is this in response to "shouldn't return true for ...", as in, you're agreeing with my reasoning and this shouldn't return true, or is that in response to "should be considered read", as in, you disagree and it shouldn't consider this as being read? Wasn't able to read it out of that sentence😅 

To make my point clearer, it shouldn't lint because:
```rs
match read() {
  Ok(bytes @ _) => println!("Read {bytes}"),
  _ => todo!()
}
```
Admittedly, it's weird code to write in the first place, but the bytes are being handled.

As for if this should be a util function, I'm not sure if it's worth. Is that useful in other lints? It seems rather niche to me to want to check if a pattern is `Ok(_)` specifically.

Reply:
Yes, I am agreeing with your reasoning :D 

As for the suggestion to be a util; there's an `is_ok` function that works only for match arms and is part of another a different function so perhaps there's use for one of the two.